src/substrate/poller.ts

```diff
@@ -65,4 +70,32 @@ export class Poller extends IEventPoller<ApiPromise, Block> {
 
     return blocks;
   }
+
+  /**
+   * Connects to chain, fetches blocks specified in given range in provided batch size,
+   * prcoesses the blocks if a handler is provided else returns the blocks for 
+   * further processing
+   * @param range IDisconnectedRange having startBlock and optional endBlock
+   * @param batchSize size of the batch in which blocks are to be fetched from chain
+   * @param processBlockFn an optional function to process the blocks
+   */
+  public async archive(range: IDisconnectedRange, batchSize: number = 500, processBlockFn: (block: Block) => any = null): Promise<Block[]> {
+    if(!range.endBlock){
+      const header = await this._api.rpc.chain.getHeader();
+      range.endBlock =  +header.number;
+    }
+    const blocks = [];
+    for (let block = range.startBlock; block < range.endBlock; block = Math.min(block + batchSize, range.endBlock)) {
+      try {
+        blocks.push(...await this.poll({startBlock: block, endBlock: Math.min(block + batchSize, range.endBlock)}, batchSize));
+        if(processBlockFn){
+          await Promise.all(blocks.map(processBlockFn));
```

Comment:
Logic error here. Because we first push each new batch onto the `blocks` array and then process all of them every iteration, we will end up calling `processBlockFn` multiple times on each batch besides the last.

Should assign each batch to a temporary variable and process them separately from the entire array of blocks.

It's also worth considering whether we should even maintain a "master" array of all blocks -- this may be quite large, causing memory overflows etc depending on the length of the chain.

Reply:
Ahh! my bad will fix the first issue. I think your concern is quite valid, I myself was a bit reluctant to store all blocks in array. At worst some one might start from block 0 till current header(which is at around 4 million atm), this can definitely cause overflow.

The way we are using the `poller.archive` method in subscribeFunc doesn't warrant us to return the array as we are not using the returned array at all. I added the master array thinking that some one in future might not want the `poller.archive` to process the blocks and would like to have all the blocks returned back and do something on that array of blocks.

Let me know how would you prefer to handle this.

Reply:
I think discarding the blocks once processed is fine for now. We can pretty easily add a full array return in the future if needed, or we can do it via a flag.